constraint/pkg/apis/templates/transform.go

```diff
@@ -0,0 +1,84 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"fmt"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+var trueBool = true
+
+// AddPreserveUnknownFields recurses through an *apiextensionsv1beta1.JSONSchemaProps
+// data structure, adding `x-kubernetes-preserve-unknown-fields: true` at every level
+// that type is equal to "object", "array", or is undefined.
+func AddPreserveUnknownFields(sch *apiextensionsv1beta1.JSONSchemaProps) error {
+	// An object can have values not described in the schema.  A blank Type could be
+	// anything, including an object, so we should treat it the same as those types.
+	// typeWithFields := sch.Type == "" || sch.Type == "object"
+	typeWithFields := sch.Type == "" || sch.Type == "object"
+
+	// Don't override an already set XPreserveUnknownFields
+	if typeWithFields && sch.XPreserveUnknownFields == nil {
+		sch.XPreserveUnknownFields = &trueBool
+	}
+
+	if sch.Properties != nil {
+		for k, v := range sch.Properties {
+			if err := AddPreserveUnknownFields(&v); err != nil {
+				return err
+			}
+
+			// As v is not a pointer, we need to set the contents of v back into the original data structure
+			sch.Properties[k] = v
+		}
+	}
+
+	// If the type is array, the schema of the array's items must be structural.  If the schema
+	// is undefined, we must add a blank one with x-kubernetes-preserve-unknown-fields to meet
+	// this structural item schema requirement.
+	schemaInvalid := sch.Items == nil || (sch.Items.Schema == nil && sch.Items.JSONSchemas == nil)
+	if sch.Type == "array" && schemaInvalid {
+		sch.Items = &apiextensionsv1beta1.JSONSchemaPropsOrArray{
+			Schema: &apiextensionsv1beta1.JSONSchemaProps{
+				XPreserveUnknownFields: &trueBool,
+			},
+		}
+
+		return nil
+	}
+
+	if sch.Items != nil {
```

Comment:
does this logic need to be inside a `sch.Type == "array"` check?

Reply:
It's possible for `sch.Items` to be defined while no `Type` is defined.  Basically the user can define _anything_ in this schema area, so we can't depend on the relationships between fields that we'd usually expect.

Reply:
I did some testing and it turns out you cannot define `items:` with any type other than `array`.  Also, if you have `type: array`, `items:` _must_ be defined or the CRD is not accepted by the API server.

That clarifies the relationship a little, but doesn't save us much in terms of code complexity AFAICT.

Reply:
This is an interesting question.  If the user previously specified `.Items` in an invalid context, was the API server throwing an error?

Reply:
What exactly do you mean by an "invalid context"?

`items: ` can be declared alongside `type: array` or with no `type` in a v1beta1 CRD schema.  Any other type (`string`, `object`, etc.) yields an error from the API server.

Reply:
By invalid context, I mean `type`.

My question here is in terms of correct behavior, if some combination of `items` and `properties`/`additionalProperties` are present, what should this do?

Reply:
Upon doing more testing, I'm seeing that the following is (unfortunately) _allowed_ in a `v1beta1 ConstraintTemplate`:

```
      validation:
        # Schema for the `parameters` field
        openAPIV3Schema:
          properties:
            foo:
              type: string
            bar:
              type: string
          items:
            type: string
```

My takeaway from this is that there is no XOR between these `properties`, `items`, `additionalProperties` expected in `v1beta1 ConstraintTemplate`.